src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -189,6 +189,84 @@ private string CreateJsonWebToken(JObject payload, SigningCredentials signingCre
+        /// <exception cref="SecurityTokenDecryptionFailedException">if the JWE was not able to be decrypted.</exception>
+        protected string DecryptToken(JsonWebToken jwtToken, TokenValidationParameters validationParameters)
+        {
+            if (jwtToken == null)
+                throw LogHelper.LogArgumentNullException(nameof(jwtToken));
+
+            if (validationParameters == null)
+                throw LogHelper.LogArgumentNullException(nameof(validationParameters));
+
+            if (string.IsNullOrEmpty(jwtToken.Enc))
+                throw LogHelper.LogExceptionMessage(new SecurityTokenException(LogHelper.FormatInvariant(TokenLogMessages.IDX10612)));
+
+            var keys = GetContentEncryptionKeys(jwtToken, validationParameters);
+
+            // keep track of exceptions thrown, keys that were tried
+            StringBuilder exceptionStrings = new StringBuilder();
+            StringBuilder keysAttempted = new StringBuilder();
+            foreach (SecurityKey key in keys)
+            {
+                var cryptoProviderFactory = validationParameters.CryptoProviderFactory ?? key.CryptoProviderFactory;
+                if (cryptoProviderFactory == null)
+                {
+                    LogHelper.LogWarning(TokenLogMessages.IDX10607, key);
+                    continue;
+                }
+
+                if (!cryptoProviderFactory.IsSupportedAlgorithm(jwtToken.Enc, key))
+                {
+                    LogHelper.LogWarning(TokenLogMessages.IDX10611, jwtToken.Enc, key);
+                    continue;
+                }
+
+                try
+                {
+                    return DecryptToken(jwtToken, cryptoProviderFactory, key);
+                }
+                catch (Exception ex)
+                {
+                    exceptionStrings.AppendLine(ex.ToString());
+                }
+
+                if (key != null)
+                    keysAttempted.AppendLine(key.ToString());
+            }
+
+            if (keysAttempted.Length > 0)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenDecryptionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10603, keysAttempted, exceptionStrings, jwtToken.EncodedToken)));
+
+            throw LogHelper.LogExceptionMessage(new SecurityTokenDecryptionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10609, jwtToken.EncodedToken)));
+        }
+
+        private string DecryptToken(JsonWebToken jwtToken, CryptoProviderFactory cryptoProviderFactory, SecurityKey key)
+        {
+            var decryptionProvider = cryptoProviderFactory.CreateAuthenticatedEncryptionProvider(key, jwtToken.Enc);
+            if (decryptionProvider == null)
+                throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(TokenLogMessages.IDX10610, key, jwtToken.Enc)));
+
+            var tokenParts = jwtToken.EncodedToken.Split('.');
```

Comment:
we need to do a controlled split here, with a maximum number of expected '.' (dots).
review what ValidateToken does.

Reply:
Makes sense. 👍